### PR TITLE
Extract response fields from response headers

### PR DESCRIPTION
**Pull request by camh-**
Extract values from response headers based in rules specified in the
`additional_bindings` section of an `HttpRule`, using the `custom`
pattern with a `kind` of `response_header`. This follows a similar
pattern spec to request headers, but the spec is used to parse response
headers rather than generate request headers.

Only singular or repeated scalar fields can be extracted, and only into
the top level of the response proto. A pattern can be used to extract a
single header to multiple fields and also require that literal text that
is not extracted also matches.

Headers are extracted after the response body is unmarshaled into the
response, so that headers can override the body.

If a header appears multiple times, each is extracted. If any fields in
the target message are not repeated, the value parsed from the last
header will be the value in the target message. If the field(s) are
repeated fields, they are all extracted in the order that the headers
appear.

### CODE REVIEW

File: httprule/httprule.go [hunk 1]
```diff
@@ -35,28 +36,191 @@ var (
...
 		}
 	}
 
-	if err := protojson.Unmarshal(b, target); err != nil {
-		return fmt.Errorf("cannot protojson unmarshal: %w", err)
+	if len(bytes.TrimSpace(b)) > 0 {
+		if err := protojson.Unmarshal(b, bodyTarget); err != nil {
+			return fmt.Errorf("cannot protojson unmarshal: %w", err)
+		}
+	}
```
Comment (juliaogris) [inline httprule/httprule.go]: I wonder if all of this should go into a `ParseResponseBody` function now.
Could also have an early return for `if len(bytes.TrimSpace(b)) == 0 { ...`

File: httprule/httprule_test.go [hunk 2]
```diff
@@ -65,32 +72,218 @@ func TestParseProtoResponseMsg2(t *testing.T) {
...
+		Pattern: &pb.HttpRule_Get{Get: "/"},
+		AdditionalBindings: []*pb.HttpRule{
+			customRule("response_header", "Notice: hello {a_string}. {a_int32} days to go"),
+			customRule("response_header", "Counter: {a_uint64}"),
+			customRule("response_header", "Ignored: {a_bytes}"),
+			customRule("header", "a: b"),
+		},
+	}
+	resp = newResponse(`{"a_bool": true, "a_int32": 105}`)
+	resp.Header.Set("notice", "hello julia. 76 days to go")
```
Comment (juliaogris) [inline httprule/httprule_test.go:198]: 🍪

**Review by juliaogris — APPROVED** [review]
Really, really neat - thank you sooo much for doing this.
🌴

File: httprule/httprule.go (hunk 1 above)
Comment (camh-) [reply httprule/httprule.go]: done

**Pull request merged**